client/session.go

```diff
@@ -835,14 +872,14 @@ func (s *session) writeAttempt(
 		return timestampErr
 	}
 
-	if s.RLock(); s.state != stateOpen {
-		s.RUnlock()
-		return errSessionStateNotOpen
+	if s.state.RLock(); s.state.status != statusOpen {
```

Comment:
This might just be personal preference and I know this was already like this, but my first instinct when I see this line is that you're using some variant of an RLock which returns immediately whether or not it was able to acquire the lock.

Might be a little easier to parse if we just move the RLock() call above the conditional

Reply:
Yeah, I'm more of a fan of idiomatic Go and advised I think when this was first authored to do what you're suggesting.  I'll clean this up now.